Rust/src/tree/point_adder.rs

```diff
@@ -0,0 +1,308 @@
+use crate::RCFFloat;
+use super::{Tree, Node, Cut, BoundingBox};
+
+/// The result of a point addition operation.
+pub enum AddResult {
+    AddedPoint(usize),
+    MassIncreased(usize),
+}
+
+
+/// A mechanism for adding points to trees.
+/// 
+/// Initialized with a mutable reference to a tree, this struct contains all of
+/// the instructions necessary to add points to trees. This code is factored
+/// out from `Tree` for improved readability.
+pub struct PointAdder<'a, T> {
+    tree: &'a mut Tree<T>,
+}
```

Comment:
Is it possible to have multiple `impl` blocks for the same struct in different files? If so, you could keep your code organization without having to introduce wrapper structs.

Reply:
Great idea. I initially misread this as "move everything into the same `impl` block" but got it right the second time. The only downside is that private functions cannot be shared between different `impl` blocks. That's probably good by design, but it means that there are a handful of helper functions that I needed to make public as I cannot create two private copies,  one for each block.